<?xml version="1.0" encoding="UTF-8"?>
<Node>
	<Permissions>
	</Permissions>
	<Doc lang="pt-PT">
	</Doc>
	<Libraries>
		<Use name="DateTimeLib" type="lib" />
		<Use name="UtilsLib" type="lib" />
	</Libraries>
	<Configuration />
	<Screens>
	</Screens>
	<Actions>
	</Actions>
	<Library forceinterface="true">
		<Function name="prepare" doc="prepare">
			<Params />
			<Return type="text" doc="Test documentation" />
			<Begin>
				<Var name="context" key="message">framework::framework.unitTest:: validate error code in capability</Var>
				<Return>
					<Var name="context" />
				</Return>
			</Begin>
		</Function>
		
		<Function name="run" doc="Run the test">
			<Params>
				<Param name="context" doc="Context data" />
			</Params>
			<Return type="single" doc="The result">
				<External name="code" type="int" doc="The result code" />
				<External name="data" type="text" doc="Data returned by the function" />
			</Return>
			<Errors>
				<Error code="-1" doc="Test failed" />
			</Errors>
			<Begin>
				<Transaction>
					<Var name="test" key="result">
						<String />
					</Var>
					<!-- Add Role -->
					<Call library="framework,permission,api" function="add">
						<Param name="code">unitTestRole</Param>
						<Param name="name">Unit test Role</Param>
						<Param name="description">Unit test Role</Param>
						<Param name="is_role">1</Param>
					</Call>

					<!-- Add capability -->
					<Call library="framework,permission,api" function="add">
						<Param name="code">unitTestCapability</Param>
						<Param name="name">Unit test Capability</Param>
						<Param name="description">Unit test Capability</Param>
						<Param name="is_role">0</Param>
					</Call>

					<!-- Add the relation -->
					<Var name="code"><Int>-1</Int></Var>
					<Var name="result" key="data">No Exception thrown</Var>
					<Try>
						<Instructions>
							<Var name="relationIdOriginal">
								<Call library="framework,permission,api" function="addRelationByCode">
									<Param name="role_code">unitTestRole</Param>
									<Param name="permission_code">unitTestCapability_WRONG</Param>
								</Call>
							</Var>						
						</Instructions>
						<Catch exception="framework/permission::invalidCode">
							<Var name="code"><Int>0</Int></Var>
							<Var name="result" key="data"><String.parse>Exception thrown $EXCEPTION->name</String.parse></Var>
						</Catch>
							<Var name="code"><Int>-1</Int></Var>
							<Var name="result" key="data">Wrong Exception</Var>
						<Catch>
						</Catch>
					</Try>

					<Var name="result" key="code"><Var name="code" /></Var>
					
					<RollBack />
				</Transaction>
				<Return>
					<Var name="result" />
				</Return>
			</Begin>
		</Function>
		
		<Function name="clean" doc="Removes the data added">
			<Params>
				<Param name="context" doc="Context data" />
			</Params>
			<Return type="int" doc="" />
			<Begin>
			</Begin>
		</Function>
	</Library>
</Node>
